### ISSUE TITLE
Easier punctuation and keyword parsing

### ISSUE THREAD

**Issue by dtolnay** [dtolnay/syn#225]
Currently parsing punctuation looks like `syn!(ShlEq)` and parsing a keyword looks like `syn!(Default_)`. I would much rather write `punct!(<<=)` and `keyword!(default)` instead -- and have that produce the correct struct type without me having to look up its name.

**Comment by dtolnay** [comment]
As part of this, let's also explore what it would look like to use type macros for token types, rather than giving them names like `ShlEq` and `Default_`. Type macros have been stable since 1.13.

```rust
ItemExternCrate {
    pub attrs: Vec<Attribute>,
    pub vis: Visibility,
    pub extern_token: token!(extern),
    pub crate_token: token!(crate),
    pub ident: Ident,
    pub rename: Option<(token!(as), Ident)>,
    pub semi_token: token!(;),
}
```

**Comment by dtolnay** [comment]
Once the user is no longer typing the token names, will probably want to rename the various token structs to prefix the name, like ShlEq -> TokenShlEq like we do for items and exprs. This will avoid conflicts with prelude types like Default and syn types like Type (after we rename Ty -> Type).

**Comment by dtolnay** [comment]
Implemented in #241.
